README
This version of fckgLite supports the Dokuwiki 'safe' file system encoding for non-Roman 
character sets.  It also supports url-style encoding, which was the encoding 
for all previous versions of Dokuwiki.

This version also supports the the jQuery conversion which Dokuwiki made beginning
with the Angua distribtution.

/*********************************************/
IMPORTANT FOR USERS OF ANTEATER
'safe' encoding was implemented in Dokuwiki's Anteater distribution, dated 2010-11-07.
A bug was found in that implementation which affected Windows users and was 
corrected in Rincewind (2011-05-25).  On an upgrade from Anteater to Rincewind,
Rincewind automatically corrects the file system for compatibility with the corrected format.

This distribution of fckgLite employs the corrected implementation of 'safe' encoding.

If you are using 'safe' encoding and are still using Anteater, you have two choices.
1. You can upgrade to Rincewind or later.
2. You can revert to the Anteater implementation of safe encoding.  
    Included in the fckg directory of this distribution is a file named revert.tgz
	Go to the fckg directory and unpack this file:
	       tar -xzf revert.tgz
	You must do this with every new install of fckgLite (or fckgLiteSafe) until
	you upgrade Dokuwiki to Rincewind or later.

/*********************************************/


See the fckgLite web site for more information about this distribution:
http://www.mturner.org/fckgLite/doku.php?id=introduction#fckglitesafe


For information on the installation and features of fckgLite, see
the fckgLite web site:  http://www.mturner.org/fckgLite


There are two versions of .htaccess files found in fckg/fckeditor/userfiles.  One of these
is called .htaccess.security and allows direct http access only to image files, but not
to any others.  The other is called .htaccess.open and allows access to any files in the
data/media directory.  In this case security is enforced using the "secure" directory
system.  

See the section on Security on the fckgLite web site:
   http://www.mturner.org/fckgLite/doku.php?id=media&#security_and_the_media_directory
Also: http://www.dokuwiki.org/plugin:fckglite?&#security


------------


Changes
rev 15-Sep_29-12_59

6 June
Upgraded from FCKeditor 2.4.1 to 2.6.6.  2.6.6 includes the scayt spellchecker, which is an in-editor spellchecker.
2.6.6 is the final FCKeditor before the change-over to CKeditor.  The upgrade required bringing the ie and gecko 
javascript files into sync, in particular so that the image and link dialogs open to a larger size.
This upgrade will now be designated fckgLite 07.

7 June 
Added configuration options for turning SCAYT on and off, setting it as auto,  and for selecting SCAYT language . 
Hide language selection button for SpellerPages spell checker when SCAYT is in effect.

9 June
Converted Arabic language file for fckg editing window from BOM to non-BOM (Corrupted image display).
Enabled overflow and scrolling in SCAYT dialog.

10 June
Increased sizes of link and image dialog windows in gecko. Adjusted font sizes in link.html

11 June 
Corrected url in Anteater safe encodig message (meta.php)

5 July 2012
Fix to FCKEditor fck_link.js for IE9, where FCKeditor's Selection.SelectNode() fails to insert link at 
cursor position if text not previously selected. This only comes into play if fckgLite is prevent from
auto-switching IE to IE8 compatibility while in editing mode.

7 July 2012
Fix to FCKeditor fck_image.js for IE9, where FCK.InsertElement failed to insert image at
cursor. (See fck_link 5 July 2012).

9 July
Removed some unnecessary files from distribution and comproessed  old Changs files to zip.

Confined IE image and link fixes to IE9 since FCKeditor will not load in IE10.  Users of IE10
where templates force Standards Mode will have to manually switch to compatibilty mode.
In other cases, fckgLite auto-switches to IE8 compatibility,

23 August
Fixed for smileys in lists where CR was lost if smiley was in li

24 August
fckeditor/dwsmileys.php: set alternate path for ajax access to smiley.conf in debian and ubuntu

25 August
Set IE compatiblity mode to IE=EmulateIE8 in file browser when browserMode > 8 (used writeln in file: fckxml.js).
This was needed by IE 10.

28 August
Implemented option to rewrite FCKeditor smiley URLs when pages are moved to new server. Regex at top of
edit.php:_preprocess.  The URLs are reset on a page by page basis when loaded and saved.  The option
should be turned off except when needed.

16 September 2012
Added experimental handling of tables, attempting to solve problems in rendering complex tables that have multiple col and row spans.
This uses a new configuration option which will turn on the experimenmtal algorithm for all pages, or it can be turned
on, on a per-page basis, either by use of a checkbox or by adding a syntax plugin style switch to the top of the page:
~~COMPLEX_TABLES~~.  Since ~~COMPLEX_TABLES~~ is implemented as a syntax plugin, it will come into effect only after the page
has been saved; before the save, the user should check off the complex tables checkbox.  The logorithm is not perfect, but it is
considerably superior to the standard table-handling parser.

17 September 2012
Minor edit to edit.php

19 September
Created loop in insert_table() that goes through all rows checking for rowspans. Rowspan property of rows array is set to 0 for
all rows where rowspans have been processed, causing them to be ignored on subsequent passes of loop.

Minor edit to edit.php editor form syntax

9 October 2012
Fixed bug where table markup inside code blocks dropped empty cells (edit.php::_preprocess)

20 Nov
Added session_id check to config.php:  in some system setting session which has already been set throws
an error.
Added $mssg="" to FileUpload() in commands.php for SendUploadResults(), without which some systems freeze.
Both of the above occurred on Windows with PHP 5.3

25
Improved mime-type handling in config.php and fckconfig.js

28 Nov
Fixed font tool plugin to work with stricter define checking of current browsers.

29 Nov
Fixed regex which removed precent-escaped backslashes in plugin syntax so that backslashes are
removed only around non-alpha characters, preserving sequences such as \r\n.

30 Nov
Added two  short-cut keys for numbered list: ctrl-+ and ctrl+{.  Both do the same thing.

8 Dec
In edit.php: insert this.is_smiley = false after alt (text form of smiley) is added to results;
prevents corruption of text formats and insertion of javascript 'undefined' when alt is not found
on next time through start()

Dec 21 2012
Updated css for new template

Dec 22
Updated fck_editorarea.css.new_template for code blocks


23 December
Replaced default fck_editorarea.css with original that was accidentally exchanged for the new

Created an IE CSS hack for geshi snippets in IE8+ and HTML5

26 December
Fix for nested lists in webkit browsers

2013 January 4
Fix for urls in DW footnotes, which were lost on re-saves.
Still-to-be-done: handling of media files in DW footnotes

Jan 4
fix to DW footnote handling to prevent url from being transferred to subsequent fotnote text
after update to previous notes's url using link tool.

Jan 5
Upgraded DW footnotes to handle links to media files

10 Jan
Added tooltip for Scayt

Jan 14
Removed "new by reference" assignment in fckg/syntax/dwplugin.php (line 191)

Jan 17
Made an exception for svg-edit svg images in php/connectors/php/util.php (l. 150), which
rejected (function: DetectHtml) them because they contained a <title> tag.

Jan 23
Upgraded Geshi plugin (input from Kamil Demecki). 

Jan 28
This version of the ctables distribution combines with and replaces the unbuntu-debian distribution.

Jan 31
In preprocessor, edit.php ca. 235ff:  Create tagged <code>. . .</code> blocks from code blocks marked
up by indentation.  This insures that the code blocks are correctly processed in the fckgLite parser.

Feb 2
convert html tag markup < and > to entities in indented code blocks
(In edit.php:_preprocess())

Feb 13
replace '>' in interwiki markup in lists, removed in above.

Removed new by reference warning  in edit.php on line 2552

Feb 14

action/meta.php
<?php
/**
 *
 */
 
if(!defined('DOKU_INC')) die();
if(!defined('DOKU_PLUGIN')) define('DOKU_PLUGIN',DOKU_INC.'lib/plugins/');
require_once(DOKU_PLUGIN.'action.php');
global $conf;
global $fckg_lang;


$default_english_file = DOKU_PLUGIN . 'fckg/action/lang/en.php';
require_once($default_english_file);
if(isset($conf['lang'])) {
  $default_lang_file = DOKU_PLUGIN . 'fckg/action/lang/' . $conf['lang'] . '.php';
  if(file_exists($default_lang_file)) {                                       
    @include_once($default_lang_file);
  }
}
 
class action_plugin_fckg_meta extends DokuWiki_Action_Plugin {
  var $session_id = false;    
  var $draft_file;
  var $dokuwiki_priority;

   function action_plugin_fckg_meta() {
      
      $this->dokuwiki_priority = $this->getConf('dw_priority');
   }
  /*
   * Register its handlers with the dokuwiki's event controller
   */
  function register(&$controller) {            
            $controller->register_hook( 'TPL_METAHEADER_OUTPUT', 'AFTER', $this, 'loadScript');    
            $controller->register_hook( 'HTML_EDITFORM_INJECTION', 'AFTER', $this, 'preprocess'); 
            $controller->register_hook( 'HTML_EDITFORM_OUTPUT', 'BEFORE', $this, 'insertFormElement');            
            $controller->register_hook('DOKUWIKI_STARTED', 'BEFORE', $this, 'file_type');         
            $controller->register_hook('TPL_CONTENT_DISPLAY', 'AFTER', $this, 'setupDWEdit');       
            $controller->register_hook('DOKUWIKI_STARTED', 'AFTER', $this, 'fnencode_check');                 
  }

 
 function  insertFormElement(&$event, $param) {	 
   global $FCKG_show_preview;  

  $param = array();

   global $ID;
   $dwedit_ns = @$this->getConf('dwedit_ns');
   if(isset($dwedit_ns) && $dwedit_ns) {
       $ns_choices = explode(',',$dwedit_ns);
       foreach($ns_choices as $ns) {
         $ns = trim($ns);
         if(preg_match("/$ns/",$ID)) {
            echo "<style type = 'text/css'>#edbtn__preview,#edbtn__save, #edbtn__save { display: inline; } </style>";         
            break;
         }
       }
   }
   $act = $event->data;
   if(is_string($act) && $act != 'edit') {  
        return;
   }

  // restore preview button if standard DW editor is in place
  // $FCKG_show_preview is set in edit.php in the register() function
if($_REQUEST['fck_preview_mode'] != 'nil' && !isset($_COOKIE['FCKG_USE']) && !$FCKG_show_preview) {    
     echo '<style type="text/css">#edbtn__preview { display:none; }</style>';
 }
 elseif($FCKG_show_preview) {
      echo '<style type="text/css">#edbtn__preview { display: inline; } </style>';
 }
 else {
    echo '<style type="text/css">#edbtn__preview, .btn_show { position:absolute; visibility:hidden; }</style>';
 }
  
 global $fckg_lang;

  if($_REQUEST['fck_preview_mode']== 'preview'){
    return;
  }

 $param = array();
 $this->preprocess($event, $param);  // create the setDWEditCookie() js function
 $button = array
        (
            '_elem' => 'button',
            'type' => 'submit',
            '_action' => 'cancel',
            'value' => $fckg_lang['btn_fck_edit'],
            'class' => 'button',
            'id' => 'edbtn__edit',            
            'title' => $fckg_lang['btn_fck_edit']             
        );

     $pos = strpos($_SERVER['HTTP_USER_AGENT'], 'MSIE');
     if($pos === false) {
                 $button['onclick'] = 'return setDWEditCookie(1, this);';
     }
     else {
                $button['onmousedown'] = 'return setDWEditCookie(1, this);';
     }

    $pos = $event->data->findElementByAttribute('type','submit');
       //inserts HTML data after that position.
    $event->data->insertElement(++$pos,$button);

   return;
 
  }


 function preprocess(&$event, $param) {	 
    $act = $event->data;
 
   if(is_string($act) && $act != 'edit') {  
        return;
   }
  global $INFO;
  $cname =  $INFO['draft'];   
  $dokuwiki_priority =$this->dokuwiki_priority;
  echo "<script type='text/javascript'>\n//<![CDATA[ \n";
  echo "var useDW_Editor =$dokuwiki_priority;";
  echo "\n //]]> </script>\n";
  echo <<<SCRIPT
    <script type="text/javascript">
    //<![CDATA[ 
    
    function setDWEditCookie(which, e) { 
       var cname = "$cname";       
       var dom = document.getElementById('fckg_mode_type');          
       if(which == 1) {
           dwedit_draft_delete("$cname");
             if(useDW_Editor) {
                document.cookie = 'FCKG_USE=other;expires=';             
              }  
           else {
                document.cookie='FCKG_USE=other;expires=Thu,01-Jan-70 00:00:01 GMT;'
           }
           if(e && e.form) {
                    if(e.form['mode']) {
                       e.form['mode'].value = 'fck';
                    }
                    else {
                       e.form['mode'] = new Object;
                       e.form['mode'].value = 'fck';  
                    }
           }
           else dom.value = 'fck';  
       }
        else {            
            document.cookie = 'FCKG_USE=_false_;expires=';                         
            dom.value = 'dwiki';        

        }
         e.form.submit();
    }
  
 
    //]]> 

    </script>
SCRIPT;

  }

function check_userfiles() {	  
 
    if($this->getConf('no_symlinks')) {	
	   return;
	}
	
    global $INFO;
    global $conf;
	$userfiles = DOKU_PLUGIN . 'fckg/fckeditor/userfiles/';
    
    $save_dir = trim($conf['savedir']);  
// msg('BASE='. DOKU_BASE);
// msg(DOKU_URL);
// msg('REL='. DOKU_REL);
    if(!preg_match('#^\.\/data$#',$save_dir)) {
        $data_media = $conf['savedir']  . '/media/';
        $domain = trim(DOKU_BASE,'/');    
        $expire = null;        
        
        if(! empty($domain )) {
        list($prefix,$mdir) = explode(trim(DOKU_BASE, '/'),$userfiles);
           $mdir = ltrim($mdir, '/');
        $media_dir = DOKU_BASE . $mdir . 'image/';
        setcookie('FCK_media',$media_dir, $expire, '/');           
        }
        else $media_dir = '/lib/plugins/ckgedit/fckeditor/userfiles/image/';        
     }
     else {
         $data_media = DOKU_INC.'data/media/';
     }
     
     if($this->getConf('winstyle')) return;    
     if(!is_writable($userfiles)){
              msg("  fckg cannot write to $userfiles. Please check the permissions.");
		      return;
     }		   
	$version = io_readFile(DOKU_PLUGIN . 'fckg/version');
	if(!$version) return;
    $meta = metaFN('fckl:symchk','.meta'); 
	$symcheck = io_readFile($meta);	    
    if($symcheck) {
	   if(trim($version)== trim($symcheck)) {  //symlinks should already have been created	 
		  return;
	   }
    }
				
	if (function_exists('php_uname')) {
	   $sys = php_uname() ;
	   if( preg_match('/Windows/i',  $sys) ) {
		     preg_match('/build\s+(\d+)/',$sys, $matches);	  
		    if($matches[1]  < 6000) {  // we can make symlinks for vista (6000) or later 
			   return;
		   }		
		 
		   $winlinks =  array();
		   $userfiles = str_replace('/', '\\',$userfiles);		   
		   exec("dir " . $userfiles, $output);
		   foreach($output as $line) {
		      if(preg_match('/<SYMLINKD>\s+(.*?)\s+\[/i',$line,$matches)) {
			     $winlinks[] = $matches[1];
			  }
		   }
		}
		
    }  
	else if( preg_match('/WINNT/i',  PHP_OS) ) {    // if we can't get php_uname and a build and this is Windows, just return
         return;
    }
	
       $show_msg = false;
	   if($INFO['isadmin'] || $INFO['ismanager'] )    {  // only admins and mgrs get messages
	       $show_msg = true;		   
	   }
	   $link_names = array('flash',  'image',  'media', 'file', 'image');
	   if(count($winlinks)) {
	       $link_names = array_diff($link_names, $winlinks);
	   }
	   $links = array();
	   foreach ($link_names as $ln) {
	        $links[$ln] = $userfiles . $ln;
	   }

      $bad_create = false; 
	  $successes =  array();
	  if(@file_exists($userfiles)) {
		   foreach($links as $name => $path) {		
			  if(!is_link($path)) {		                      
                     if(file_exists($path) && is_file($path) ){
					       unlink($path);
                       }					 
                    if(file_exists($path) && is_dir($path) ){
					       rmdir($path);
                       }					 
					 if(!@symlink($data_media,$path) ) {
					     $bad_create = true;
						  if($show_msg)  msg("unable to create $name link:  $path",-1);			  
				   }
				   else {
				     $successes[] = $name; 
				   }
			 }	   	   
		  }
      }
	  else {
	     if($show_msg)  {
			msg("Cannot create symlinks for filebrowser.  Cannot access:  $userfiles   ",-1);
		 }
	  }
	   
	 
	   
	  if($bad_create) {
	       if($show_msg)  {
			   msg("There was an error when trying to create symbolic links in $userfiles. "
					. "See fckg/auto_install.pdf  or  the <a href='http://www.mturner.org/fckgLite/doku.php?id=docs:auto_install'>fckgLite web site</a>" , 2);					
				}
      }
	  else {	        
	       if(count($successes)) {
				$links_created = implode(', ',$successes);
				msg('The following links were created in the userfiles directory: ' . $links_created,2);
			 }
	  }
	  			io_saveFile($meta,$version);
}
            
  
  function set_session() {	
      global $USERINFO, $INFO; 
      global $conf; 
      global $ID;
      global $ACT;

      if($this->session_id) return;       

           $cname = getCacheName($INFO['client'].$ID,'.draft');  
           $fckl_draft = $cname . '.fckl';
           if((isset($ACT) && is_array($ACT)) || isset($_REQUEST['dwedit_preview'])) {
              if(isset($ACT['draftdel']) || isset($ACT['cancel']) || isset($_REQUEST['dwedit_preview'])) {
                 @unlink($fckl_draft);   
                 @unlink($cname); 
              }
           }

           if(file_exists($cname)) {
              if(file_exists($fckl_draft)) {
                    unlink($fckl_draft);
              }
              @rename($cname, $fckl_draft);
           }

          
           $session_string =  session_id(); 
           $this->session_id = $session_string;      
       

           $_SESSION['dwfck_id'] = $session_string; 
           $default_fb = $this->getConf('default_fb');

           if($default_fb == 'none') {
               $acl = 255; 
           }
           else {
              $acl = auth_quickaclcheck($ID); 
           }
           $_SESSION['dwfck_acl'] = $acl; 

           if($this->getConf('openfb') || $acl == 255) {
             $_SESSION['dwfck_openfb'] = 'y';
           }
           else { 
              $_SESSION['dwfck_openfb'] = 'n';
           }

           $_SESSION['dwfck_grps'] = $USERINFO['grps'];
           $_SESSION['dwfck_client'] = $INFO['client'];   
           $_SESSION['dwfck_sepchar'] = $conf['sepchar'] ;   
           $_SESSION['dwfck_conf'] = array('sepchar'=> $conf['sepchar'],
                  'isadmin'=>($INFO['isadmin'] || $INFO['ismanager']), 
                  'deaccent'=>$conf['deaccent'], 'useslash'=>$conf['useslash']);
           $elems = explode(':', $ID);  
           array_pop($elems);
 
           $_SESSION['dwfck_ns'] = implode(':',$elems);        
           $_SESSION['dwfck_top'] = implode('/',$elems);           
           $_SESSION['dwfck_del'] = $this->getConf('acl_del');
           
            // temp fix for puzzling encoding=url bug in frmresourceslist.html,
           // where image loading is processed in GetFileRowHtml()

           if(preg_match('/fckg:fckeditor:userfiles:image/',$ID)) {
                      $_SESSION['dwfck_ns'] = "";        
                      $_SESSION['dwfck_top'] = "";      

            }

          // $expire = time()+60*60*24*30;
           $expire = null;
           setcookie('FCK_NmSp_acl',$session_string, $expire, '/');           

           setcookie('FCK_SCAYT',$this->getConf('scayt'), $expire, '/');                
           setcookie('FCK_SCAYT_AUTO',$this->getConf('scayt_auto'), $expire, '/'); 
           $scayt_lang = $this->getConf('scayt_lang');
           if(isset($scayt_lang)) {
               list($scayt_lang_title,$scayt_lang_code) = explode('/',$scayt_lang);
               if($scayt_lang_code!="en_US") {
                  setcookie('FCK_SCAYT_LANG',$scayt_lang_code, $expire, '/'); 
               }
           }
           
            if ($this->getConf('winstyle')) {
              setcookie('FCKConnector','WIN', $expire, DOKU_BASE);                                
            }  
            
           if ($this->dokuwiki_priority) {
               if(isset($_COOKIE['FCKG_USE']) && $_COOKIE['FCKG_USE'] == 'other') {                              
                   $expire = time() -60*60*24*30;
                   setcookie('FCKG_USE','_false_', $expire, '/');           
               }
               else {            
                  setcookie('FCKG_USE','_false_', $expire, '/');           
                }
           }
                   
  }

  function file_type(&$event, $param) {	 
       global $ACT, $TEXT;
       global $USERINFO, $INFO, $ID; 

	   $this->check_userfiles(); 
	   
       if(isset($_COOKIE['FCK_NmSp'])) $this->set_session(); 
       /* set cookie to pass namespace to FCKeditor's media dialog */
      // $expire = time()+60*60*24*30;
       $expire = null;
       setcookie ('FCK_NmSp',$ID, $expire, '/');     
      
          

      /* Remove TopLevel cookie */         
       if(isset($_COOKIE['TopLevel'])) {
            setcookie("TopLevel", $_REQUEST['TopLevel'], time()-3600, '/');
       }

     
       if(!isset($_REQUEST['id']) || isset($ACT['preview'])) return;
       if(isset($_REQUEST['do']) && isset($_REQUEST['do']['edit'])) {
              $_REQUEST['do'] = 'edit';
       }
  } 

function loadScript(&$event) {
  echo <<<SCRIPT

    <script type="text/javascript">
    //<![CDATA[ 
    function LoadScript( url )
    {
     document.write( '<scr' + 'ipt type="text/javascript" src="' + url + '"><\/scr' + 'ipt>' ) ;        

    }
//]]> 

 </script>

SCRIPT;

}

/** 
 *  Handle features need for DW Edit: 
 *    1. load script, if not loaded
 *    2. Re-label Cancel Button "Exit" when doing a preview  
 *    3. set up $REQUEST value to identify a preview when in DW Edit , used in 
 *       set_session to remove fckgLite and DW drafts if present after a DW preview  
*/
  function setupDWEdit(&$event) {
  global $ACT;

  $url = DOKU_BASE . 'lib/plugins/fckg/scripts/script-cmpr.js';
  echo <<<SCRIPT

    <script type="text/javascript">
    //<![CDATA[ 

    try {
    if(!HTMLParserInstalled){
         LoadScript("$url");        
    }
    }
    catch (ex) {  
         LoadScript("$url");        
    }             

    function createRequestValue() {
        try{
        var inputNode=document.createElement('input');
        inputNode.setAttribute('type','hidden');
        inputNode.setAttribute('value','yes');
        inputNode.setAttribute('name','dwedit_preview');
        inputNode.setAttribute('id','dwedit_preview');
        var dwform = GetE("dw__editform");
        dwform.appendChild(inputNode);
        }catch(e) { alert(e); }
    }
//]]> 
 </script>

SCRIPT;

  if(isset($_REQUEST['do']) && is_array($_REQUEST['do'])) {
    if(isset($_REQUEST['do']['preview'])) {
           echo '<script type="text/javascript">';
           echo ' var dwform = GetE("dw__editform"); dwform["do[draftdel]"].value = "Exit";';
           echo "\ncreateRequestValue()\n";
           echo  '</script>';
    }
  }


  }

function is_safeUpgraded() {
   $safescript = DOKU_PLUGIN . 'fckg/scripts/safeFN_class.js';
    if(!file_exists($safescript) ){	    
		return false; 
	}
   $lines = file($safescript );
	
	for($i=0; $i<count($lines); $i++) {    
	  if(strpos($lines[$i],'/**')){
		 if(isset($lines[$i+1])) { 
			if(stripos($lines[$i+1], 'upgrade') !== false) {
			 return true;
		  } 
		}
	   }
	  }
	  return false;
}
  
function fnencode_check() {


       global $conf;
       global $updateVersion;
       $rencode = false;
	  
        if($conf['fnencode'] != 'safe') return;

        if(isset($updateVersion) && $updateVersion >= 31) {           
          $rencode = true;     
        }
        else {
            $list = plugin_list('action');
            if(in_array('safefnrecode', $list)){
                $rencode = true;   
     
            }
            elseif(file_exists($conf['datadir'].'_safefn.recoded') ||
               file_exists($conf['metadir'].'_safefn.recoded') ||
               file_exists($conf['mediadir'].'_safefn.recoded') )
            { 
               $rencode = true;
            }
        }


      if($rencode && !file_exists(DOKU_PLUGIN . 'fckg/saferencode')) {
         msg("This version of fckgLite does not support the re-encoded safe filenames. "
         . "You risk corrupting your file system.  Download an fnrencode version from either gitHub or the fckgLite web site."
         . " <a style='color:blue' href='http://www.mturner.org/fckgLite/doku.php?id=docs:upgrade_6&#anteater'>or the fckgLite web site</a>  ",
            -1);
      }	 
      else if(!$rencode && file_exists(DOKU_PLUGIN . 'fckg/saferencode') && $this->is_safeUpgraded() )   {	    
	    msg("This version of fckgLiteSafe requires a newer version of Dokuwiki (2011-05-25 Rincewind or later).  You risk corrupting your file system. "
		 .   "To convert this distribution of fckgLite/fckgLiteSafe for use with earlier versions of Dokuwiki,  see the README file or " 
		 . " <a style='color:blue' href='http://www.mturner.org/fckgLite/doku.php?id=docs:upgrade_6&#anteater'>or the fckgLite web site</a>  ",
		-1);
    }	  
}




      


function write_debug($data) {
  return;
  if (!$handle = fopen(DOKU_INC .'meta.txt', 'a')) {
    return;
    }
  if(is_array($data)) {
     $data = print_r($data,true);
  }
    // Write $somecontent to our opened file.
    fwrite($handle, "$data\n");
    fclose($handle);

}

}



